submariner-operator/diagnose-refactor-ocm.md

```diff
@@ -0,0 +1,151 @@
+ This is to reduce complexity on OCM UI side to pass credentials for remote cluster which are required for Firewall option.
+ This support will be added in later release.
+
+### Alternate Design
+
+* Instead of `--output json` option to provide results in json format for SubmarinerAddon to consume, it will just retain
+ existing text output.
+* UI will provide a link to logs of `submariner-diagnose` Pod and show results as is to the user.
+* Advantage of this design is that newer features can be added to `subctl diagnose` without requiring corresponding changes
+ in UI to consume those results.
+* This will also simplify the code in SubmarinerAddon as it will no longer need to extract and parse diagnose results.
+
+### SubmarinerDiagnose CRD
+
+```Go
+// SubmarinerDiagnoseSpec defines the desired configuration to run SubmarinerDiagnose
+type SubmarinerDiagnoseSpec struct {
+All             bool         `json:"all,omitempty"`
+CNI             bool         `json:"CNI,omitempty"`
+Connections     bool         `json:"connections,omitempty"`
+Deployment      bool         `json:"deployment,omitempty"`
+Firewall        bool         `json:"firewall,omitempty"`
+GatherLogs      bool         `json:"gatherLogs,omitempty"`
+K8sVersion      bool         `json:"k8sVersion,omitempty"`
+KubeProxyMode   bool         `json:"kubeProxyMode,omitempty"`
+FirewallOptions FirewallOptions `json:"firewallOptions,omitempty"`
+
+// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+// Important: Run "make manifests" to regenerate code after modifying this file
+// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+}
+
+type FirewallOptions struct {
+InterCluster bool `json:"interCluster,omitempty"`
+IntraCluster bool `json:"intraCluster,omitempty"`
+Metrics      bool `json:"metrics,omitempty"`
+RemoteCluster            string `json:"remoteCluster,omitempty"`
+RemoteK8sAPIServer       string `json:"remoteK8sAPIServer,omitempty"`
+RemoteK8sAPIServerToken  string `json:"remoteK8sAPIServerToken,omitempty"`
+RemoteK8sCA              string `json:"remoteK8sCA,omitempty"`
+RemoteK8sSecret          string `json:"remoteK8sSecret,omitempty"`
+RemoteK8sRemoteNamespace string `json:"remoteK8sRemoteNamespace,omitempty"`
+}
+
+// SubmarinerDiagnoseStatus defines the observed result of SubmarinerDiagnose
+type SubmarinerDiagnoseStatus struct {
+Conditions []metav1.Condition `json:"conditions"`
+K8sVersion string `json:"k8sVersion,omitempty"`
+CNIType string `json:"cniType,omitempty"`
+KubeProxyMode bool `json:"kubeProxyMode,omitempty"`
+ConnectionsStatus ConnectionsStatus `json:"connectionsStatus,omitempty"`
+FirewallStatus FirewallStatus `json:"firewallStatus,omitempty"`
+// DeploymentStatus already captured in SubmarinerStatus, no need to duplicate information.
+// DeploymentStatus SubmarinerStatus `json:"DeploymentStatus,omitempty"`
+}
+
+type ConnectionsStatus struct {
+// TBD
```

Comment:
Still TBD?

Reply:
Had discussions with UI team, so we're dropping this for now. We'll have to revisit this when UI supports topoolgy view. At that time we will define these in an easy to consume format for the UI.